### ISSUE TITLE
On auto-save cursor jumps to the bottom

### ISSUE THREAD

**Issue by Hossman333** [reasonml-editor/atom-ide-reason#39]
Every time I autosave and refmt my code it jumps to the bottom of the file.

Atom editor version: 1.33.1
`$ refmt --version` gives me Reason 3.3.3

This also happens on atom-beta for me. 

Is it because i'm using vim-mode-plus? Quite obnoxious so definitely would like to fix it 😅😊

**Comment by alex35mil** [comment]
Apparently, formatting fix for OLS caused this behavior in RLS. That's sad b/c it means that to make both work, OLS and RLS integrations must be separate packages since provided services configuration is done on `package.json` level. /cc @zaaack

**Comment by zaaack** [comment]
@Hossman333 @alexfedoseev I fixed this via diffing new text and old text(adopted from [reason-refmt](https://github.com/reasonml-old/atom-reason-refmt/blob/master/lib/main.js#L114)) which should be done on language servers, also I added a formatOnSave option so we can disable this and use reason-refmt. A new patch version is published.

But then @alexfedoseev 's comment reminded me of OLS, sadly it throws an error here, and I have no clue what caused this. Does `formatOnSave` on OLS work for you before? 

```sh
logger.ts:23 Reason (reason) rpc.sendRequest textDocument/formatting threw Error: Request textDocument/formatting failed with message: "file" argument must be a non-empty string
    at new ResponseError (/Users/z/Projects/Atom/atom-ide-reason/node_modules/vscode-jsonrpc/lib/messages.js:46)
    at handleResponse (/Users/z/Projects/Atom/atom-ide-reason/node_modules/vscode-jsonrpc/lib/main.js:430)
    at processMessageQueue (/Users/z/Projects/Atom/atom-ide-reason/node_modules/vscode-jsonrpc/lib/main.js:258)
    at Immediate._onImmediate (/Users/z/Projects/Atom/atom-ide-reason/node_modules/vscode-jsonrpc/lib/main.js:242)
    at runCallback (timers.js:789)
    at tryOnImmediate (timers.js:751)
    at processImmediate [as _immediateCallback] (timers.js:722)
    at Function.formatDocument (code-format-adapter.ts:75)
    at ReasonMLLanguageClient.<anonymous> (auto-languageclient.ts:669)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/z/Projects/Atom/atom-ide-reason/node_modules/atom-languageclient/build/lib/auto-languageclient.js:4)
    at <anonymous>
```

**Comment by alex35mil** [comment]
@zaaack OLS's format on save works for me in current version w/ the fix that caused cursor jumping on RLS (I switched back to OLS b/c RLS stopped working in my app, it hangs on almost every change and loads 100% CPU).

**Comment by zaaack** [comment]
@alexfedoseev Have you tried to restart all servers after switching? I have to say the fix is not fast but cannot imagine it's so slow...

**Comment by alex35mil** [comment]
@zaaack Hmmm, `1.1.4` doesn't work for me: neither RLS nor OLS.
RLS gives a bunch of odd errors in console on save: https://cl.ly/1963f7225ff4
OLS can't even initialize.

**Comment by zaaack** [comment]
@alexfedoseev  It's wired...  I don't think these errors are related, hoping @Hossman333 will have some good news when he gives feedback.

**Comment by alex35mil** [comment]
@zaaack I re-checked. You're right, errors are unrelated, but formatting still doesn't work. I checked on few projects using published `1.1.4` and my local `1.1.3` (w/ `--dev` flag):
- OLS can't initialize w/ `1.1.4`, works fine w/ local `1.1.3`
- RLS doesn't do anything on save w/ `1.1.4`, and has jumping cursor issue w/ local `1.1.3`

**Comment by zaaack** [comment]
@alexfedoseev You can try with master branch, I think I fixed RLS, OLS can initialize on my here, but cannot format on save, do you use the newest atom?

**Comment by alex35mil** [comment]
Yep, I'm on latest. Confirmed that RLS works on master 👍

**Comment by Hossman333** [comment]
Sorry about the delayed response! Wow, thank you for the quick responses and solutions! The latest published version `1.1.4` fixed the issue. (: 

Thank you for your time and work!

**Comment by alex35mil** [comment]
@zaaack One more issue w/ current `master`: `Cmd + Z` seems broken. When save event is fired and then I hit undo -> things get messed up.

TBH I incline more to not messing w/ our own re-implementations of language client/server features b/c it might be a can of worms...

**Comment by alex35mil** [comment]
Also, it looks like 2+ empty lines are not collapsing to 1 anymore.

EDIT: Nope, it’s b/c sometimes it takes a significant time to apply formatting.

**Comment by zaaack** [comment]
@alexfedoseev Thanks for reporting!

> One more issue w/ current master: Cmd + Z seems broken. When save event is fired and then I hit undo -> things get messed up.

I think this can be fixed via 881cfcbb9e7..

> TBH I incline more to not messing w/ our own re-implementations of language client/server features b/c it might be a can of worms...

Well, I don't want to re-implemente language client/server features, too. But if we want to implement a usable format-on-save feature, we need `reason-language-server` to provide a minimal-change edit list, which avoids the cursor jumps to the bottom, and the `atom-languageclient` to fix issues of conflicts with `insert final new line` feature(this can be fixed via RLS keep the `final new line` behavior when formatting or atom-languageclient considering `final new line` when applying edits), that's why I applied edits on my own... 

Before upstream repositories fix these issues, I think there are three ways for us to choose. The first one is using current implementation as a temporary resolution; The second one is disabled this feature and comment out code; or, keep current implementation, but disable this feature by default and add some warning, let users decide whether to take the risks.
